README.md
# StructyRecord

`StructyRecord` provides a Struct-like interface for your `Record`s.

- Use your record's macros in the _same module_ where it is defined!
- Access and update fields in your record through named macro calls.
- Create and update records at runtime (not limited to compile time).
- Calculate 1-based indexes to access record fields in `:ets` tables.

## Setup

The package is [available in Hex](https://hex.pm/packages/structy_record) and can
be installed by adding `structy_record` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:structy_record, "~> 0.2.1"}
  ]
end
```

## Usage

To get started, see the [documentation] for `StructyRecord.defrecord/3`:

```elixir
iex> h StructyRecord.defrecord
```

[documentation]: https://hexdocs.pm/structy_record/StructyRecord.html

### Examples

Activate this macro in your environment:

```elixir
require StructyRecord
```

Define a structy record for a rectangle:

```elixir
StructyRecord.defrecord Rectangle, [:width, :height] do
  def area(r = record()) do
    get_width(r) * get_height(r)
  end

  def perimeter(record(width: w, height: h)) do
    2 * (w + h)
  end

  def square?(record(width: same, height: same)), do: true
  def square?(_), do: false
end
```

Activate its macros in your environment:

```elixir
use Rectangle
```

Create instances of your structy record:

```elixir
rect = Rectangle.{}                        #-> {Rectangle, nil, nil}
rect = Rectangle.{[]}                      #-> {Rectangle, nil, nil}
no_h = Rectangle.{[width: 1]}              #-> {Rectangle, 1, nil}
no_w = Rectangle.{[height: 2]}             #-> {Rectangle, nil, 2}
wide = Rectangle.{[width: 10, height: 5]}  #-> {Rectangle, 10, 5}
tall = Rectangle.{[width: 4, height: 25]}  #-> {Rectangle, 4, 25}
even = Rectangle.{[width: 10, height: 10]} #-> {Rectangle, 10, 10}
```

Inspect the contents of those instances:

```elixir
rect |> Rectangle.inspect() #-> "Rectangle.{[width: nil, height: nil]}"
no_h |> Rectangle.inspect() #-> "Rectangle.{[width: 1, height: nil]}"
no_w |> Rectangle.inspect() #-> "Rectangle.{[width: nil, height: 2]}"
wide |> Rectangle.inspect() #-> "Rectangle.{[width: 10, height: 5]}"
tall |> Rectangle.inspect() #-> "Rectangle.{[width: 4, height: 25]}"
even |> Rectangle.inspect() #-> "Rectangle.{[width: 10, height: 10]}"
```

Get values of fields in those instances:

```elixir
Rectangle.{{tall, :height}}       #-> 25
Rectangle.{[height: h]} = tall; h #-> 25
tall |> Rectangle.get_height()    #-> 25
```

Set values of fields in those instances:

```elixir
Rectangle.{{even, width: 1}}    #-> {Rectangle, 1, 10}
even |> Rectangle.put(width: 1) #-> {Rectangle, 1, 10}
even |> Rectangle.put_width(1)  #-> {Rectangle, 1, 10}

Rectangle.{{even, width: 1, height: 2}}                   #-> {Rectangle, 1, 2}
even |> Rectangle.put(width: 1, height: 2)                #-> {Rectangle, 1, 2}
even |> Rectangle.put_width(1) |> Rectangle.put_height(2) #-> {Rectangle, 1, 2}
```

Use your custom code on those instances:

```elixir
rect |> Rectangle.area() #-> (ArithmeticError) bad argument in arithmetic expression: nil * nil
no_h |> Rectangle.area() #-> (ArithmeticError) bad argument in arithmetic expression: 1 * nil
no_w |> Rectangle.area() #-> (ArithmeticError) bad argument in arithmetic expression: nil * 2
wide |> Rectangle.area() #-> 50
tall |> Rectangle.area() #-> 100
even |> Rectangle.area() #-> 100

rect |> Rectangle.perimeter() #-> (ArithmeticError) bad argument in arithmetic expression: nil + nil
no_h |> Rectangle.perimeter() #-> (ArithmeticError) bad argument in arithmetic expression: 1 + nil
no_w |> Rectangle.perimeter() #-> (ArithmeticError) bad argument in arithmetic expression: nil + 2
wide |> Rectangle.perimeter() #-> 30
tall |> Rectangle.perimeter() #-> 58
even |> Rectangle.perimeter() #-> 40

rect |> Rectangle.square?() #-> true
no_h |> Rectangle.square?() #-> false
no_w |> Rectangle.square?() #-> false
wide |> Rectangle.square?() #-> false
tall |> Rectangle.square?() #-> false
even |> Rectangle.square?() #-> true
```

## License

[Spare A Life]: https://sunaku.github.io/vegan-for-life.html
> Like my work? 👍 Please [spare a life] today as thanks! 🐄🐖🐑🐔🐣🐟✨🙊✌  
> Why? For 💕 ethics, the 🌎 environment, and 💪 health; see link above. 🙇

(the ISC license)

Copyright 2019 Suraj N. Kurapati <https://github.com/sunaku>

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


CHANGELOG.md
## Version 0.2.1 (2021-02-20)

### Housekeeping:

  * README: `runtime: false` suggestion broke releases.

    The `runtime: false` flag in the user's `mix.exs` file prevents the
    `structy_record` application (which bundles module `StructyRecord`)
    from being included in app releases.  This breaks defined macros and
    functions that delegate to the `StructyRecord.from_list/3` function
    because the `StructyRecord` module won't be included in app releases!

## Version 0.2.0 (2021-02-18)

This release adds a convenient new shorthand syntax for the `record()` macro,
renames field accessors to prevent name collisions, clarifies docs, and more.

### Incompatible:

  * Rename `record!/1` function to `from_list/1`.

  * Rename `record!/2` function to `merge/2`.

  * Add `get_` and `put_` prefix to field accessors.

    Field names can no longer conflict with defined macros & functions.

  * Don't check argument types in Elixiry interface.

    It broke simple macro expansion when used in case/function clauses.

### Enhancements:

  * Add `Module.{_}` syntax for `Module.record(_)`.

    https://stackoverflow.com/a/51313720/120075

  * Add `inspect/2` for friendlier inspection.

  * Add `to_list/0` to get record's template.

  * Add `to_list/1` alias for `record/1` macro.

  * Add `index/1` to get field index in tuple.

  * Add `get/2` macro as an alias to `record/2`.

  * Add `put/2` macro as an alias to `record/2`.

  * Define documentation for all macros and functions.

### Housekeeping:

  * `record?/1` macro: only use pattern matching check.

  * `keypos/1` macro: don't call module being defined.

  * mix.exs: drop application(); use `runtime: false`.

## Version 0.1.0 (2019-10-26)

Initial release! 🎉 Announced on [ElixirForum] and [Reddit].

[ElixirForum]: https://elixirforum.com/t/26367
[Reddit]: https://www.reddit.com/r/elixir/comments/dni8zb


test/test_helper.exs
ExUnit.start()

defmodule TestHelper do
  require StructyRecord

  StructyRecord.defrecord(NoFields, [])
  StructyRecord.defrecord(OneField, [:one])
  StructyRecord.defrecord(OneFieldWithDefaultValue, one: 1)
  StructyRecord.defrecord(TwoFields, [:one, :two])

  StructyRecord.defrecord NoFieldsWithCustomDoBlock, [] do
    defmacro macro(arg) do
      quote do
        unquote(arg)
      end
    end

    def function(arg), do: arg

    def function_using_macros, do: function_using_macros(record())
    def function_using_macros(r = record()), do: r
  end
end


test/structy_record_test.exs
alias TestHelper.NoFields
alias TestHelper.NoFieldsWithCustomDoBlock
alias TestHelper.OneField
alias TestHelper.OneFieldWithDefaultValue
alias TestHelper.TwoFields

defmodule StructyRecordTest do
  use ExUnit.Case
  doctest StructyRecord

  describe "defrecord/2" do
    test "the do..end block is optional" do
      StructyRecord.defrecord(OptionalDoBlockTest, [])
    end
  end

  describe "defrecord/3" do
    test "do_block: function" do
      assert NoFieldsWithCustomDoBlock.function(:ok) == :ok
    end

    test "do_block: macro" do
      require NoFieldsWithCustomDoBlock
      assert NoFieldsWithCustomDoBlock.macro(:ok) == :ok
    end

    test "do_block: function using record() macros" do
      require Record
      result = NoFieldsWithCustomDoBlock.function_using_macros()
      assert Record.is_record(result, NoFieldsWithCustomDoBlock)
    end
  end

  describe "index/3" do
    test "empty: field not found" do
      assert_raise ArgumentError, "record Foobar does not have the key: :field", fn ->
        StructyRecord.index(:field, Foobar, [])
      end
    end

    test "single: field is found" do
      assert StructyRecord.index(:field, Foobar, field: nil) == 1
    end

    test "double: field is found" do
      assert StructyRecord.index(:field, Foobar, other_field: nil, field: nil) == 2
    end
  end

  describe "from_list/2" do
    test "empty" do
      assert [] |> StructyRecord.from_list(Foobar) == {Foobar}
    end

    test "single" do
      assert [one: 1] |> StructyRecord.from_list(Foobar) == {Foobar, 1}
    end

    test "double" do
      assert [one: 1, two: 2] |> StructyRecord.from_list(Foobar) == {Foobar, 1, 2}
    end
  end

  describe "from_list/3" do
    test "empty" do
      assert [] |> StructyRecord.from_list(Foobar, []) == {Foobar}
    end

    test "empty contents: retain template" do
      template = [known: :default_value]
      contents = []
      assert contents |> StructyRecord.from_list(Foobar, template) == {Foobar, :default_value}
    end

    test "empty template: ignore contents" do
      template = []
      contents = [known: :content_value]
      assert contents |> StructyRecord.from_list(Foobar, template) == {Foobar}
    end

    test "intersection: overwrite template value with content value" do
      template = [known: :default_value]
      contents = [known: :content_value]
      assert contents |> StructyRecord.from_list(Foobar, template) == {Foobar, :content_value}
    end

    test "difference: ignore content values that aren't in template" do
      template = [known: :default_value]
      contents = [known: :content_value, unknown_extra_contents: nil]
      assert contents |> StructyRecord.from_list(Foobar, template) == {Foobar, :content_value}
    end
  end

  describe "inspect/3" do
    test "empty" do
      assert [] |> StructyRecord.inspect(Foobar) == "Foobar.{}"
    end

    test "single" do
      assert [one: 1] |> StructyRecord.inspect(Foobar) == "Foobar.{[one: 1]}"
    end

    test "supports Kernel.inspect/2 options" do
      assert [one: 1, two: 2] |> StructyRecord.inspect(Foobar, limit: 1) ==
               "Foobar.{[one: 1, ...]}"
    end

    test "invalid: contents must be a list" do
      assert_raise FunctionClauseError, fn ->
        StructyRecord.inspect("string", Foobar)
      end
    end
  end

  ## record_primitives

  use NoFields
  use OneField
  use OneFieldWithDefaultValue
  use TwoFields

  describe "{}/0" do
    test "to create a new record with default values for all fields" do
      require Record
      assert NoFields.{} |> elem(0) == NoFields
      assert NoFields.{} |> Record.is_record(NoFields)
      assert NoFields.{} == NoFields.StructyRecord.record()
    end
  end

  describe "{}/1" do
    test "to create a new record with the given fields and values" do
      assert OneField.{[one: 1]} == OneField.StructyRecord.record(one: 1)
    end

    test "to get the zero-based index of the given field in a record" do
      assert OneField.{:one} == OneField.StructyRecord.record(:one)
    end
  end

  describe "{{}}/1" do
    test "to convert the given record to a keyword list" do
      record = OneField.{}
      assert OneField.{{record}} == OneField.StructyRecord.record(record)
    end
  end

  describe "{{}}/2" do
    test "to access a given field in a given record" do
      record = OneField.{}
      assert OneField.{{record, :one}} == nil

      record = OneFieldWithDefaultValue.{}
      assert OneFieldWithDefaultValue.{{record, :one}} == 1
    end

    test "to update an existing record with the given fields and values" do
      record = OneField.{}
      updated_record = OneField.{{record, one: 1}}
      assert OneField.{{updated_record, :one}} == 1
    end
  end

  describe "record/0" do
    test "to create a new record with default values for all fields" do
      require Record
      assert NoFields.record() |> elem(0) == NoFields
      assert NoFields.record() |> Record.is_record(NoFields)
      assert NoFields.record() == NoFields.StructyRecord.record()
    end
  end

  describe "record/1" do
    test "to create a new record with the given fields and values" do
      assert OneField.record(one: 1) == OneField.StructyRecord.record(one: 1)
    end

    test "to get the zero-based index of the given field in a record" do
      assert OneField.record(:one) == OneField.StructyRecord.record(:one)
    end

    test "to convert the given record to a keyword list" do
      record = OneField.record()
      assert OneField.record(record) == OneField.StructyRecord.record(record)
    end
  end

  describe "record/2" do
    test "to access a given field in a given record" do
      record = OneField.{}
      assert record |> OneField.record(:one) == nil

      record = OneFieldWithDefaultValue.{}
      assert record |> OneFieldWithDefaultValue.record(:one) == 1
    end

    test "to update an existing record with the given fields and values" do
      record = OneField.{}
      updated_record = record |> OneField.record(one: 1)
      assert updated_record |> OneField.record(:one) == 1
    end
  end

  describe "is_record/1" do
    test "checks if argument _loosely_ matches the shape of this record" do
      assert NoFields.{} |> NoFields.is_record()
      assert NoFields.record() |> NoFields.is_record()
      assert NoFields.StructyRecord.record() |> NoFields.is_record()
      assert {NoFields, :extra_field_is_NOT_checked_by_guard} |> NoFields.is_record()
    end
  end

  describe "record?/1" do
    test "checks if argument _strictly_ matches the shape of this record" do
      assert NoFields.{} |> NoFields.record?()
      assert NoFields.record() |> NoFields.record?()
      assert NoFields.StructyRecord.record() |> NoFields.record?()

      refute {NoFields, :any_extra_field_is_also_checked_by_this_macro}
             |> NoFields.record?()
    end
  end

  ## elixiry_interface

  describe "index/1" do
    test "to get the zero-based index of the given field in a record" do
      assert OneField.index(:one) == OneField.StructyRecord.record(:one)
    end

    test "support runtime evaluation via StructyRecord.index() fallback" do
      runtime = :one
      assert OneField.index(runtime) == OneField.StructyRecord.record(:one)
    end
  end

  describe "keypos/1" do
    test "to get the 1-based index of the given field in a record" do
      assert OneField.keypos(:one) == 1 + OneField.StructyRecord.record(:one)
    end
  end

  describe "get/2" do
    test "to fetch the value of a given field in a given record" do
      record = OneField.{}
      assert record |> OneField.get(:one) == nil
    end
  end

  describe "put/2" do
    test "to update an existing record with the given fields and values" do
      record = OneField.{}
      updated_record = record |> OneField.put(one: 1)
      assert updated_record |> OneField.get(:one) == 1
    end
  end

  describe "from_list/1" do
    test "to create a new record at runtime with the given fields and values" do
      test_runtime_record_creation_from_collection(Keyword.new())
      test_runtime_record_creation_from_collection(Map.new())
    end

    defp test_runtime_record_creation_from_collection(collection) do
      contents = [one: 1] |> Enum.into(collection)
      assert OneField.from_list(contents) == OneField.{[one: 1]}
      assert OneField.from_list([]) == OneField.{}
    end

    test "uses default field value when no value is specified in new contents" do
      contents = [two: 2]
      runtime_record = OneFieldWithDefaultValue.from_list(contents)
      assert runtime_record |> OneFieldWithDefaultValue.get_one() == 1
    end
  end

  describe "merge/2" do
    test "accepts a record as input" do
      record = OneField.{}
      other = OneField.{[one: "other"]}
      merged = record |> OneField.merge(other)
      assert merged == other
    end

    test "accepts collections as input" do
      test_runtime_record_merging_from_collection(Keyword.new())
      test_runtime_record_merging_from_collection(Map.new())
    end

    defp test_runtime_record_merging_from_collection(collection) do
      record = OneField.{}
      assert record |> OneField.get_one() == nil

      contents = [one: 1] |> Enum.into(collection)
      runtime_record = record |> OneField.merge(contents)
      assert runtime_record |> OneField.get_one() == 1
    end
  end

  describe "to_list/0" do
    test "to get a template of fields and default values for this record" do
      assert NoFields.to_list() == []
      assert OneField.to_list() == [one: nil]
      assert OneFieldWithDefaultValue.to_list() == [one: 1]
    end
  end

  describe "to_list/1" do
    test "to convert a record into a Keyword list" do
      record = NoFields.{}
      assert NoFields.to_list(record) == []

      record = OneFieldWithDefaultValue.{}
      assert OneFieldWithDefaultValue.to_list(record) == [one: 1]
    end
  end

  describe "inspect/2" do
    test "to pretty-print a record with its field names and values" do
      record = NoFields.{}
      module = inspect(NoFields)
      assert NoFields.inspect(record) == "#{module}.{}"

      record = OneField.{[one: 1]}
      assert OneField.inspect(record) == "#{inspect(OneField)}.{[one: 1]}"
    end
  end

  ## field_accessors

  describe "get_${field}/1" do
    test "to access a specific field in a given record" do
      record = OneField.{}
      assert record |> OneField.get_one() == nil

      record = OneFieldWithDefaultValue.{}
      assert record |> OneFieldWithDefaultValue.get_one() == 1
    end
  end

  describe "put_${field}/2" do
    test "to assign a specific field in a given record" do
      record = OneField.{}
      assert record |> OneField.put_one(1) == OneField.{[one: 1]}
    end
  end
end


mix.exs
defmodule StructyRecord.MixProject do
  use Mix.Project

  @version "0.2.1"
  @github "https://github.com/sunaku/structy_record"

  def project do
    [
      app: :structy_record,
      version: @version,
      elixir: "~> 1.6",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      docs: docs(),
      package: package(),
      name: "StructyRecord",
      description: "Provides a Struct-like interface for your Records."
    ]
  end

  defp deps do
    [
      {:ex_doc, "~> 0.21", only: :dev, runtime: false}
    ]
  end

  defp docs do
    [
      source_ref: "v#{@version}",
      source_url: @github,
      extras: ["README.md", "CHANGELOG.md"],
      main: "readme"
    ]
  end

  defp package do
    [
      licenses: ["ISC"],
      links: %{"GitHub" => @github}
    ]
  end
end


lib/structy_record.ex
defmodule StructyRecord do
  @moduledoc """

  `StructyRecord` provides a Struct-like interface for your `Record`s.

  - Use your record's macros in the _same module_ where it is defined!
  - Access and update fields in your record through named macro calls.
  - Create and update records at runtime (not limited to compile time).
  - Calculate 1-based indexes to access record fields in `:ets` tables.

  """

  @doc """
  Defines a module named `alias` that is also a `Record` composed of `fields`.

  ## Parameters

  - `alias` is the name of the module being defined.  It also serves as the
  `tag` parameter of `Record.defrecord/3`, which helps identify the record.

  - `fields` specifies the shape of the record being defined.  It is either:
    - a list of `Atom` field names whose default values are always `nil`
    - a `Keyword` list of field names along with their own default values

  - `do_block` is an optional block of code that is passed into `defmodule/2`.
  It allows you to extend the module being defined with your own custom code,
  which has compile-time access to all the guards and macros described below.

  ## Results

  The defined module contains the following guards, macros, and functions.

  Guards (available at compile time):
  - `is_record/1` to check if argument _loosely_ matches this record's shape

  Macros (available at compile time):
  - `{}/0` to create a new record with default values for all fields
  - `{}/1` to create a new record with the given fields and values
  - `{}/1` to get the zero-based index of the given field in a record
  - `{{}}/1` to convert a record into a list of its fields and values
  - `{{}}/2` to get the value of a given field in a given record
  - `{{}}/2` to assign the given fields and values in a given record
  - `record?/1` to check if argument _strictly_ matches this record's shape
  - `record/0` to create a new record with default values for all fields
  - `record/1` to create a new record with the given fields and values
  - `record/1` to get the zero-based index of the given field in a record
  - `record/1` to convert a record into a list of its fields and values
  - `record/2` to get the value of a given field in a given record
  - `record/2` to assign the given fields and values in a given record
  - `get/2` to fetch the value of a given field in a given record
  - `put/2` to assign the given fields and values inside a given record
  - `get_${field}/1` to fetch the value of a specific field in a given record
  - `put_${field}/2` to assign the value of a specific field in a given record
  - `index/1` to get the zero-based index of the given field in a record
  - `keypos/1` to get the 1-based index of the given field in a record
  - `to_list/0` to get a template of fields and default values for this record
  - `to_list/1` to convert a record into a list of its fields and values

  Functions (available at runtime only):
  - `from_list/1` to create a new record with the given fields and values
  - `merge/2` to assign the given fields and values inside a given record
  - `inspect/2` to inspect the contents of a record using `Kernel.inspect/2`

  ## Examples

  Activate this macro in your environment:

      require StructyRecord

  Define a structy record for a rectangle:

      StructyRecord.defrecord Rectangle, [:width, :height] do
        def area(r = record()) do
          get_width(r) * get_height(r)
        end

        def perimeter(record(width: w, height: h)) do
          2 * (w + h)
        end

        def square?(record(width: same, height: same)), do: true
        def square?(_), do: false
      end

  Activate its macros in your environment:

      use Rectangle

  Create instances of your structy record:

      rect = Rectangle.{}                        #-> {Rectangle, nil, nil}
      rect = Rectangle.{[]}                      #-> {Rectangle, nil, nil}
      no_h = Rectangle.{[width: 1]}              #-> {Rectangle, 1, nil}
      no_w = Rectangle.{[height: 2]}             #-> {Rectangle, nil, 2}
      wide = Rectangle.{[width: 10, height: 5]}  #-> {Rectangle, 10, 5}
      tall = Rectangle.{[width: 4, height: 25]}  #-> {Rectangle, 4, 25}
      even = Rectangle.{[width: 10, height: 10]} #-> {Rectangle, 10, 10}

  Inspect the contents of those instances:

      rect |> Rectangle.inspect() #-> "Rectangle.{[width: nil, height: nil]}"
      no_h |> Rectangle.inspect() #-> "Rectangle.{[width: 1, height: nil]}"
      no_w |> Rectangle.inspect() #-> "Rectangle.{[width: nil, height: 2]}"
      wide |> Rectangle.inspect() #-> "Rectangle.{[width: 10, height: 5]}"
      tall |> Rectangle.inspect() #-> "Rectangle.{[width: 4, height: 25]}"
      even |> Rectangle.inspect() #-> "Rectangle.{[width: 10, height: 10]}"

  Get values of fields in those instances:

      Rectangle.{{tall, :height}}       #-> 25
      Rectangle.{[height: h]} = tall; h #-> 25
      tall |> Rectangle.get_height()    #-> 25

  Set values of fields in those instances:

      Rectangle.{{even, width: 1}}    #-> {Rectangle, 1, 10}
      even |> Rectangle.put(width: 1) #-> {Rectangle, 1, 10}
      even |> Rectangle.put_width(1)  #-> {Rectangle, 1, 10}

      Rectangle.{{even, width: 1, height: 2}}                   #-> {Rectangle, 1, 2}
      even |> Rectangle.put(width: 1, height: 2)                #-> {Rectangle, 1, 2}
      even |> Rectangle.put_width(1) |> Rectangle.put_height(2) #-> {Rectangle, 1, 2}

  Use your custom code on those instances:

      rect |> Rectangle.area() #-> (ArithmeticError) bad argument in arithmetic expression: nil * nil
      no_h |> Rectangle.area() #-> (ArithmeticError) bad argument in arithmetic expression: 1 * nil
      no_w |> Rectangle.area() #-> (ArithmeticError) bad argument in arithmetic expression: nil * 2
      wide |> Rectangle.area() #-> 50
      tall |> Rectangle.area() #-> 100
      even |> Rectangle.area() #-> 100

      rect |> Rectangle.perimeter() #-> (ArithmeticError) bad argument in arithmetic expression: nil + nil
      no_h |> Rectangle.perimeter() #-> (ArithmeticError) bad argument in arithmetic expression: 1 + nil
      no_w |> Rectangle.perimeter() #-> (ArithmeticError) bad argument in arithmetic expression: nil + 2
      wide |> Rectangle.perimeter() #-> 30
      tall |> Rectangle.perimeter() #-> 58
      even |> Rectangle.perimeter() #-> 40

      rect |> Rectangle.square?() #-> true
      no_h |> Rectangle.square?() #-> false
      no_w |> Rectangle.square?() #-> false
      wide |> Rectangle.square?() #-> false
      tall |> Rectangle.square?() #-> false
      even |> Rectangle.square?() #-> true

  """
  defmacro defrecord(alias, fields, do_block \\ []) do
    definition = alias |> append_alias(:StructyRecord)
    field_names = fields |> field_names()

    quote do
      require Record, as: StructyRecord_Record

      defmodule unquote(definition) do
        @moduledoc false
        StructyRecord_Record.defrecord(:record, unquote(alias), unquote(fields))
      end

      defmodule unquote(alias) do
        require unquote(definition), as: StructyRecord_Definition
        alias __MODULE__, as: StructyRecord_Interface

        defmacro __using__(_options) do
          quote do
            require StructyRecord_Definition
            require StructyRecord_Interface
          end
        end

        unquote(record_primitives())
        unquote(elixiry_interface())
        unquote(field_accessors(field_names))
        unquote(do_block)
      end
    end
  end

  defp append_alias({tag = :__aliases__, context, namespace}, addendum) do
    {tag, context, namespace ++ [addendum]}
  end

  defp record_primitives do
    quote do
      @doc """
      Creates a new record with fields set to default values.
      """
      defmacro {} do
        quote do
          StructyRecord_Definition.record()
        end
      end

      @doc """
      Either fetches the value of a given field in a given record,
      or assigns the given fields and values inside a given record.
      """
      defmacro {{record, field_or_contents}} do
        quote do
          StructyRecord_Definition.record(unquote(record), unquote(field_or_contents))
        end
      end

      @doc """
      Converts the given record into a list of its fields and values.
      """
      defmacro {{_tag = :{}, _context, [record]}} do
        quote do
          StructyRecord_Definition.record(unquote(record))
        end
      end

      @doc """
      Either creates a new record with the given fields and values,
      or returns the zero-based index of the given field in a record.
      """
      defmacro {contents_or_field_or_record} do
        quote do
          StructyRecord_Definition.record(unquote(contents_or_field_or_record))
        end
      end

      @doc """
      Either creates a new record with the given fields and values,
      or returns the zero-based index of the given field in a record,
      or converts the given record into a list of its fields and values.
      Defaults to creating a new record with fields set to default values.
      """
      defmacro record(contents_or_field_or_record \\ []) do
        quote do
          StructyRecord_Definition.record(unquote(contents_or_field_or_record))
        end
      end

      @doc """
      Either fetches the value of a given field in a given record,
      or assigns the given fields and values inside a given record.
      """
      defmacro record(record, field_or_contents) do
        quote do
          StructyRecord_Definition.record(unquote(record), unquote(field_or_contents))
        end
      end

      @doc """
      Checks if the given argument _loosely_ matches this record's shape.
      """
      defguard is_record(record)
               when record
                    |> StructyRecord_Record.is_record(StructyRecord_Interface)

      @doc """
      Checks if the given argument _strictly_ matches this record's shape.
      """
      defmacro record?(record) do
        quote do
          match?(StructyRecord_Definition.record(), unquote(record))
        end
      end
    end
  end

  defp elixiry_interface do
    quote do
      @record StructyRecord_Definition.record()
      @template StructyRecord_Definition.record(@record)

      @doc """
      Returns the zero-based index of the given field in this kind of record.
      """
      defmacro index(field) when is_atom(field) do
        quote do
          StructyRecord_Definition.record(unquote(field))
        end
      end

      defmacro index(field) do
        quote bind_quoted: [field: field, template: @template] do
          StructyRecord.index(field, StructyRecord_Interface, template)
        end
      end

      @doc """
      Returns the 1-based position of the given field in this kind of record.
      """
      defmacro keypos(field) do
        quote do
          1 + StructyRecord_Definition.record(unquote(field))
        end
      end

      @doc """
      Returns a template of fields and default values for this kind of record.
      """
      defmacro to_list do
        quote do
          unquote(@template)
        end
      end

      @doc """
      Converts the given record into a `Keyword` list of its fields and values.
      """
      defmacro to_list(record) do
        quote do
          StructyRecord_Definition.record(unquote(record))
        end
      end

      @doc """
      Fetches the value of the given field in the given record.
      """
      defmacro get(record, field) do
        quote do
          StructyRecord_Definition.record(unquote(record), unquote(field))
        end
      end

      @doc """
      Assigns the given fields and values inside a given record.
      """
      defmacro put(record, contents) do
        quote do
          StructyRecord_Definition.record(unquote(record), unquote(contents))
        end
      end

      @doc """
      Creates a new record _at runtime_ with the given fields and values.
      """
      def from_list(contents) do
        StructyRecord.from_list(contents, StructyRecord_Interface, @template)
      end

      @doc """
      Assigns the given fields and values _at runtime_ inside a given record.
      """
      def merge(record, contents_or_record)

      def merge(record, other = record()), do: merge(record, to_list(other))

      def merge(record, contents) do
        template = record |> StructyRecord_Definition.record()
        StructyRecord.from_list(contents, StructyRecord_Interface, template)
      end

      @doc """
      Inspects the given record's contents _at runtime_ using `Kernel.inspect/2`.
      """
      def inspect(record, options \\ []) do
        contents = record |> StructyRecord_Definition.record()
        StructyRecord.inspect(contents, StructyRecord_Interface, options)
      end
    end
  end

  defp field_accessors(field_names) do
    getters = field_names |> Enum.map(&getter_macro/1)
    putters = field_names |> Enum.map(&putter_macro/1)

    quote do
      unquote(getters)
      unquote(putters)
    end
  end

  defp field_names(fields) do
    fields
    |> Enum.map(fn
      {field, _default_value} -> field
      field -> field
    end)
  end

  defp getter_macro(field) do
    quote do
      @doc """
      Fetches the value of the `#{unquote(inspect(field))}` field in the given record.
      """
      defmacro unquote(:"get_#{field}")(record) do
        quote do
          StructyRecord_Definition.record(unquote(record), :field)
        end
        |> case do
          {call, meta, _args = [record, :field]} ->
            args = [record, unquote(field)]
            {call, meta, args}
        end
      end
    end
  end

  defp putter_macro(field) do
    quote do
      @doc """
      Assigns the value of the `#{unquote(inspect(field))}` field in the given record.
      """
      defmacro unquote(:"put_#{field}")(record, value) do
        quote do
          StructyRecord_Definition.record(unquote(record), unquote(value))
        end
        |> case do
          {call, meta, _args = [record, value]} ->
            args = [record, [{unquote(field), value}]]
            {call, meta, args}
        end
      end
    end
  end

  @doc false
  # Returns the zero-based index of the given field in the given kind of record.
  def index(field, record_tag, record_template) do
    case find_index(record_template, field) do
      nil ->
        # error out in _exactly_ the same way as Record.index/3 for uniformity
        raise ArgumentError,
              "record #{inspect(record_tag)} does not have the key: #{inspect(field)}"

      index ->
        # add +1 for record_tag which occupies the first position in the tuple
        1 + index
    end
  end

  defp find_index(template, field) do
    template
    |> Enum.find_index(&match?({^field, _default_value}, &1))
  end

  @doc false
  # Creates a new record of the given type with the given fields and values.
  def from_list(contents, record_tag) do
    values = Keyword.values(contents)
    [record_tag | values] |> :erlang.list_to_tuple()
  end

  @doc false
  # Creates a new record of the given type with the given fields and values
  # according to the given template of known fields and their default values.
  def from_list(contents, record_tag, record_template) do
    contents
    |> intersect(record_template)
    |> from_list(record_tag)
  end

  defp intersect(contents, template) do
    template
    |> Enum.map(fn {field, template_value} ->
      value = Access.get(contents, field, template_value)
      {field, value}
    end)
  end

  @doc false
  # Inspects the contents of the given record type using `Kernel.inspect/2`.
  def inspect(contents, record_tag, options \\ []) when is_list(contents) do
    "#{inspect(record_tag)}.{#{inspect_contents(contents, options)}}"
  end

  defp inspect_contents([], _), do: ""
  defp inspect_contents(list, options), do: Kernel.inspect(list, options)
end
